### ISSUE TITLE
EventPublisher consumes 100% CPU

### ISSUE THREAD

**Issue by white-hat** [saltstack/salt#35503]
### Description of Issue/Question

EventPublisher process starts to consume 100% CPU, minions become mostly unavailable and any state is applied very slowly and unreliably.
Usually EventPublisher is idling, never saw it consuming more then 1-2% CPU.
I do not know what triggers this issue, but it's a floating bug, usually happens on the 7-15th day of uptime.
### Setup

hw: 8 cores, 32g ram, LA ~ 0.50
tcp transport enabled
~500 minions
Running strace against the stuck proccess shows:

```
sendto(289, "\202\244body\332\0\313salt/job/20160815162022"..., 218, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(394, "\202\244body\332\0\313salt/job/20160816103806"..., 218, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(109, "\202\244body\332\0\303salt/job/20160813080233"..., 210, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(130, "\202\244body\332\0\303salt/job/20160813081334"..., 210, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(212, "\202\244body\332\2\24salt/auth\n\n\205\242id\246hss485\246"..., 547, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(321, "\202\244body\332\2/salt/auth\n\n\205\242id\332\0!devop"..., 574, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(128, "\202\244body\332\0\313salt/job/20160813080843"..., 218, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(389, "\202\244body\332\2\26salt/auth\n\n\205\242id\247hss1062"..., 549, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(344, "\202\244body\332\0\306salt/job/20160816013507"..., 213, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(217, "\202\244body\332\0\303salt/job/20160814192519"..., 210, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(364, "\202\244body\332\0\303salt/job/20160816040259"..., 210, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(140, "\202\244body\332\0\303salt/job/20160813134021"..., 210, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
sendto(314, "\202\244body\332$\227salt/job/20160815210650"..., 9382, 0, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
```
### Steps to Reproduce Issue

Run salt-master for 7-15 days
### Versions Report

```
# salt-call --versions-report
Salt Version:
           Salt: 2016.3.2

Dependency Versions:
           cffi: 1.5.2
       cherrypy: 2.3.0
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 2.0.6
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.12 (default, Jul  1 2016, 15:12:24)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: 0.2.4
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
        machine: x86_64
        release: 4.4.0-31-generic
         system: Linux
        version: Ubuntu 16.04 xenial

# pip freeze
apache-libcloud==0.20.0
azure==1.0.3
azure-common==1.1.4
azure-mgmt==0.20.2
azure-mgmt-common==0.20.0
azure-mgmt-compute==0.20.1
azure-mgmt-network==0.20.1
azure-mgmt-nspkg==1.0.0
azure-mgmt-resource==0.20.1
azure-mgmt-storage==0.20.0
azure-nspkg==1.0.0
azure-servicebus==0.20.1
azure-servicemanagement-legacy==0.20.2
azure-storage==0.20.3
cffi==1.5.2
chardet==2.3.0
CherryPy==2.3.0
croniter==0.3.8
cryptography==1.2.3
enum34==1.1.2
futures==3.0.5
gitdb==0.6.4
GitPython==2.0.6
idna==2.0
inotify==0.2.6
ipaddress==1.0.16
Jinja2==2.8
lockfile==0.12.2
Mako==1.0.3
MarkupSafe==0.23
msgpack-python==0.4.6
ndg-httpsclient==0.4.0
netaddr==0.7.18
ply==3.7
pyasn1==0.1.9
pycparser==2.14
pycrypto==2.6.1
pycurl==7.43.0
PyMySQL==0.7.2
pyOpenSSL==0.15.1
python-apt==1.1.0b1
python-dateutil==2.4.2
python-systemd==231
pytz==2014.10
PyYAML==3.11
pyzmq==15.2.0
requests==2.9.1
salt==2016.3.2
setproctitle==1.1.10
simplejson==3.8.1
six==1.10.0
smmap==0.9.0
timelib==0.2.4
tornado==4.2.1
ujson==1.35
urllib3==1.13.1
```

**Comment by thatch45** [comment]
Thanks @white-hat !
This looks like we are either not covering a case where the the socket becomes unavailable or this is part of a buffer overrun error on the socket.
Can you send us your master config? I am particularly wondering if you have changed the `sock_dir` or `root_dir` values. Also, what do your filesystem mounts look like? I wonder if s tmp process could be attempting to remove the long standing socket.

**Comment by cachedout** [comment]
This smells a little like ephemeral port exhaustion. Most systems will throw an error like this if that happens.

**Comment by white-hat** [comment]
So it's stuck again... strace shows same errors, I've checked the number of open socket FDs:
```
# ls -l /proc/6169/fd/ |grep socket| wc -l
442
```
and 40 after salt-master restart
What else can I do next time this happens?

**Comment by white-hat** [comment]
> I am particularly wondering if you have changed the sock_dir or root_dir values
no I have not
I use single root partition, but I use gitfs

**Comment by Ch3LL** [comment]
@white-hat would you mind also sharing a sanitized version of your master? Also anything unique about yoru environment such as syndics or multi-masters? trying to determine if this is similar to #36285

**Comment by white-hat** [comment]
@Ch3LL 
We have a single master, we use events, mine, external pillars and TCP(tornado) tranport

**Comment by Ch3LL** [comment]
@white-hat there are a couple of issues mentioned in #36285 which might help in your scenario.

1. Depending on your master config you might be running into this issue:  #35480
2. A known issue that might be related which is fixed in #36024 
3. There is also another PR that might help your situation here: #36184

Would you mind giving the PRs a chance and see if they help out?

**Comment by MaxBosse** [comment]
Hey,
We were having a similar issue (EventPublisher consuming 100% CPU) when doing some states that would run on all servers (~700, 2 syndics, batches of ~20) but the EventPublisher would go down in CPU again after the states were fully run.

I just upgraded our salt-master to 2016.3.3 and it seemed to fix this issue for now. I am running 3 jobs simultaneously (each one would have pushed the CPU to 100% before) and I am not even able to see the EventPublisher in top. I will let you know if I see issues pop up again after the salt-masters ran for a few days but for now it looks good.

```
Salt Version:
           Salt: 2016.3.3
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: 3.2.3
       dateutil: 2.5.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.24.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.12 (default, Aug 31 2016, 12:37:48)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.1.5
 
System Versions:
           dist:   
        machine: amd64
        release: 10.1-RELEASE-p16
         system: FreeBSD
        version: Not Installed
```

**Comment by Ch3LL** [comment]
@MaxBosse thanks for the input... @white-hat if you upgrade to 2016.3.3 do you see any improvements?

**Comment by cachedout** [comment]
@MaxBosse Could you provide an update here? Has this situation been resolved with 2016.3.3?

**Comment by cachedout** [comment]
There has no been reply here so I am closing this issue.
